Forces gases to obey certain laws of nature

DaedalusGame opened the pull request:
:cl:
fix: Fixes Noblium Formation being multiplicative, so having 500 moles of tritium and 500 moles of nitrogen no longer produces 2500 moles of noblium (it should correctly produce 10 moles)
fix: Fixes most reactions deleting more gas than exists and making gas out of nowhere
fix: Fixes Stim Formation invoking a byond bug and not using its intended polynomial formula
fix: Fixes Cryo Cells not having garbage_collect and clamping
fix: Fixes Rad Collectors not having clamping
fix: Fixes Division by Zero when Fusion has no impurities.
fix: Removes a redundant line in lungs
fix: Fixes rad_act signal not firing from turfs, living mobs, rad collectors and geiger counters
/:cl:

I need someone to look over my math again because I'm not 100% sure this is right.
Things that should probably be done before merging this:
- ~~Fire, but it's complicated~~
- ~~Fusion, but it's disabled (Can we get a rework already damnit)~~
- Add clamping to everything that removes gases because obviously we can't trust garbage_collect()

Did you know that TRITIUM_BURN_TRIT_FACTOR is a constant factor for how insane tritium is?
Closes #33532

DaedalusGame committed:
- Fixes most gas reactions deleting more gas than exists and thus making gas out of nowhere.
- Makes Noblium formation no longer multiplicative.
- Expunges invocation of a byond bug by stimulum production.

DaedalusGame committed:
- Redundant parenthesis
- I'm going back to sleep

DaedalusGame committed:
- Sends fire and fusion to jail

DaedalusGame committed:
- Deletes redundant line in lung code
- Adds garbage_collect and clamp to cryocells
- Adds clamp to collectors

DaedalusGame committed:
- Reverted the tritium burn changes because apparently I'm dumb

DaedalusGame committed:
- FUCK

DaedalusGame committed:
- Fixes division by zero when fusion has no impurities

vuonojenmustaturska commented on code/modules/atmospherics/gasmixtures/reactions.dm:
```diff
@@ -207,13 +208,15 @@
 	var/plasma_fused = (PLASMA_FUSED_COEFFICENT*catalyst_efficency)*(temperature/PLASMA_BINDING_ENERGY)/10
 	var/tritium_catalyzed = (CATALYST_COEFFICENT*catalyst_efficency)*(temperature/PLASMA_BINDING_ENERGY)/40
+	plasma_fused = min(plasma_fused,cached_gases[/datum/gas/plasma][MOLES])
+	tritium_catalyzed = min(tritium_catalyzed,cached_gases[/datum/gas/tritium][MOLES])
```
please combine these two assignments with the var initialization above

duncathan commented on code/game/turfs/open.dm:
```diff
@@ -353,6 +353,7 @@
 
 /turf/open/rad_act(pulse_strength)
 	if (air.gases[/datum/gas/carbon_dioxide] && air.gases[/datum/gas/oxygen])
+		pulse_strength = min(pulse_strength,air.gases[/datum/gas/carbon_dioxide][MOLES]*1000,air.gases[/datum/gas/oxygen][MOLES]*2000)
```
could you leave a comment on these and similar limits elsewhere in the PR indicating that this is ensuring that matter is conserved appropriately?

duncathan commented on code/modules/atmospherics/gasmixtures/reactions.dm:
```diff
@@ -252,6 +255,7 @@
 
 	var/old_heat_capacity = air.heat_capacity()
 	var/heat_efficency = temperature/(FIRE_MINIMUM_TEMPERATURE_TO_EXIST*100)
+	heat_efficency = min(heat_efficency,cached_gases[/datum/gas/oxygen][MOLES],cached_gases[/datum/gas/nitrogen][MOLES])
```
as naksu said above

duncathan commented on code/modules/atmospherics/gasmixtures/reactions.dm:
```diff
@@ -284,6 +288,7 @@
 
 	var/old_heat_capacity = air.heat_capacity()
 	var/reaction_efficency = 1/((pressure/(0.1*ONE_ATMOSPHERE))*(max(cached_gases[/datum/gas/plasma][MOLES]/cached_gases[/datum/gas/tritium][MOLES],1)))
+	reaction_efficency = min(reaction_efficency,cached_gases[/datum/gas/tritium][MOLES],cached_gases[/datum/gas/plasma][MOLES]/2)
```
^

duncathan commented on code/modules/atmospherics/gasmixtures/reactions.dm:
```diff
@@ -316,7 +321,8 @@
 	var/heat_scale = air.temperature/STIMULUM_HEAT_SCALE
 	var/stim_energy_change
-	stim_energy_change =heat_scale + (STIMULUM_FIRST_RISE(heat_scale**2)) - (STIMULUM_FIRST_DROP(heat_scale**3)) + (STIMULUM_SECOND_RISE(heat_scale**4)) - (STIMULUM_ABSOLUTE_DROP(heat_scale**5))
+	heat_scale = min(heat_scale,cached_gases[/datum/gas/tritium][MOLES],cached_gases[/datum/gas/plasma][MOLES],cached_gases[/datum/gas/nitryl][MOLES])
```
^

duncathan commented on code/modules/atmospherics/gasmixtures/reactions.dm:
```diff
@@ -345,7 +351,8 @@
 	var/old_heat_capacity = air.heat_capacity()
-	var/nob_formed = (cached_gases[/datum/gas/nitrogen][MOLES]*cached_gases[/datum/gas/tritium][MOLES])/100
+	var/nob_formed = (cached_gases[/datum/gas/nitrogen][MOLES]+cached_gases[/datum/gas/tritium][MOLES])/100
+	nob_formed = min(nob_formed,cached_gases[/datum/gas/tritium][MOLES]/10,cached_gases[/datum/gas/nitrogen][MOLES]/20)
```
^

duncathan reviewed (changes requested):
doing these in one line is ugly but unfortunately necessary for the minor performance enhancements that come with it

DaedalusGame committed:
- Adds extra commentary for the next unlucky sod
- Compresses lines for performance or so they say
- Adds ..() to rad_act on turfs

DaedalusGame committed:
- Fuck it, fixes rad collectors, open turfs, geiger counters and living mobs not signalling rad_act components
